### ISSUE TITLE
Specific remote follow stuck in "awaiting approval" state; cancel button doesn't work

### ISSUE THREAD

**Issue by confluence** [mastodon/mastodon#19645]
### Steps to reproduce the problem

I have an account on mastodon.social, and I was trying to follow a user on mastodon.online. I have successfully followed other users on multiple different servers.

Steps that I followed:

1. Logged into my mastodon.social account
2. Navigated to a user page on mastodon.online
3. Clicked the "Follow" button
4. Got a popup dialog explaining how to follow the account (it appears that this server does not provide a "remote follow" function which would allow me to follow the account directly from this user page)
5. Entering the URL provided in this dialog into the search bar on mastodon.social did not find any search results. However, I tried the @username@servername string instead, and that did work.
6. I clicked through to the search result on mastodon.social
7. I clicked on the "follow" button.

I am omitting a lot of server timeouts from the steps described above -- I understand that mastodon.social has been under heavy load today, and it has been failing for me intermittently. Unfortunately since this happened several hours ago I can't recall exactly at which points of this process timeouts happened.

### Expected behaviour

I should have started following the user.

### Actual behaviour

My request to follow the user is stuck "awaiting approval", and I cannot cancel it and try again.

### Detailed description

The follow request never completed. When I navigate to this user now (on mastodon.social, using the process described above), I see a "withdraw follow request" button (with a mouseover label explaining that the request is awaiting approval, and that I can cancel the request by clicking the button). But when I click the button, nothing happens -- there is no visible response at all. The request has been in this state for hours.

The person I tried to follow can't see any notifications about approving the request in their account. Their account is not locked. They have another follower from mastodon.social, and did not have to take any steps to approve them.

This was the first account on mastodon.online that I tried to follow, but I just tried to follow a different user on the same server using the exact same process, and it worked immediately. But that first follow request seems to be stuck in this indeterminate state, and it doesn't seem possible for me to reset it and try to follow the user again.

### Specifications

Mastodon v4.0.0rc1 (both servers)
Firefox 105.0
Ubuntu 20.04

**Comment by confluence** [comment]
At some point since I filed this the problem solved itself -- I am now following the account. So possibly this issue can be closed.

**Comment by ClearlyClaire** [comment]
> Got a popup dialog explaining how to follow the account (it appears that this server does not provide a "remote follow" function which would allow me to follow the account directly from this user page)

The old “remote follow” flow has been discontinued with the recent logged-out UI redesign, with what you experienced being the replacement. It should be functionally equivalent and is supposed to be more generic.

> Entering the URL provided in this dialog into the search bar on mastodon.social did not find any search results. However, I tried the @username@servername string instead, and that did work.

This part likely failed because mastodon.social was overloaded.

> My request to follow the user is stuck "awaiting approval"

The request being stuck on “awaiting approval” is not surprising with mastodon.social's load, which at the time likely had hours of backlog to go through.

> and I cannot cancel it and try again.

That however is pretty surprising to me if I understand correctly. What happened when you tried to cancel the follow request? Did it do nothing? Was there any error?

**Comment by confluence** [comment]
> The request being stuck on “awaiting approval” is not surprising with mastodon.social's load, which at the time likely had hours of backlog to go through.

But a subsequent follow request succeeded immediately, while the earlier request was still "stuck". In general, all the other follows I attempted (while the server was under heavy load) succeeded immediately, although I sometimes got errors and had to retry. This was the only time that something like this happened.

> That however is pretty surprising to me if I understand correctly. What happened when you tried to cancel the follow request? Did it do nothing? Was there any error?

There was no error -- the button was completely unresponsive. I checked the console in the browser inspector, but didn't see anything. I gather that there's supposed to be a confirmation dialog when you click on the button, but I never got that. I was getting repeated gateway errors (as hovering popup notifications), but it wasn't clear whether those were directly triggered by attempts to push the button or other actions on the page (as everything was extremely laggy). I also tried to trigger the cancellation by clicking on the hourglass which appeared next to the user in the search results (as per tooltip instructions), but that didn't have any effect either.

**Comment by confluence** [comment]
Update: it's happening again, this time with a user on pleroma.debian.social. Exact same symptoms (waiting for approval; cancel buttons do nothing), but this time it doesn't look like it's a load problem, since mastodon.social just got more server resources and is much more responsive.

I don't know whether this is related, but I was previously unable to follow this user at all when using the old "remote follow" mechanism from the user's page. After finding out in the earlier reply that following from inside your own server's UI was the new and preferred way of following accounts, I tried doing that instead. At first it appeared to work -- when I clicked the follow icon in the search results, it immediately changed to the "following" status. But when I looked for the user in my "following" list, they weren't there, and when I searched for them again I saw the hourglass status.

I will see if the problem resolves itself again.

**Comment by confluence** [comment]
Yes, it did. So the delay is just temporary and following eventually works. But the interface doesn't seem to be responding as expected. The next time this happens I'll see if cancellation works from Chrome/ium and from the Android app, to check whether this is Firefox-specific.

**Comment by ClearlyClaire** [comment]
The UI optimistically reflects follows as immediately accepted when the remote account is not marked as private, but a message from the remote server is still required to finalize the follow relationship, that's while when either server is under load, refreshing the page may seem to change the follow to a follow request. Such occurrences should resolve over time, though, so the initial case you are describing may be something else.

> There was no error -- the button was completely unresponsive. I checked the console in the browser inspector, but didn't see anything. I gather that there's supposed to be a confirmation dialog when you click on the button, but I never got that.

This is dependent on a specific setting that you may not have enabled: “Show confirmation dialog before unfollowing someone”. But thanks to your description, I am able to confirm that the issue indeed occurs when the setting is *disabled*, which was not the case for me, and why I could not reproduce the issue until now.

**Comment by confluence** [comment]
> Such occurrences should resolve over time, though, so the initial case you are describing may be something else.

To be clear, the initial case I reported *did* resolve over time (I reported this in my first followup comment).

> This is dependent on a specific setting that you may not have enabled: “Show confirmation dialog before unfollowing someone”. But thanks to your description, I am able to confirm that the issue indeed occurs when the setting is _disabled_, which was not the case for me, and why I could not reproduce the issue until now.

Ah, interesting. It wasn't clear to me that this setting also applied to withdrawing a follow request. But if it does, then I would expect clicking on the cancel button to have the immediate effect of cancelling the request (bypassing the confirmation dialog). Instead, there seems to be no effect on the status of the request at all -- it remains in the same waiting state until it eventually succeeds.

(The only reason I was trying to use the cancel button in the first place was that I thought that the request could have been left in an inconsistent state because the server was overloaded, and that it might be helpful to cancel it and retry. So I don't actually *need* to use the button in this situation. But this does look like a UI bug.)

**Comment by ClearlyClaire** [comment]
> > Such occurrences should resolve over time, though, so the initial case you are describing may be something else.
> 
> To be clear, the initial case I reported _did_ resolve over time (I reported this in my first followup comment).

Oh sorry, I missed that. Then it was indeed a load issue. As for the inability to cancel the follow request, it's a UI bug, already reported as #19569. Closing.
